zkevm-circuits/src/evm_circuit/execution/memory_copy.rs

```diff
@@ -0,0 +1,532 @@
+                cb.tx_lookup(
+                    tx_id.expr(),
+                    TxContextFieldTag::CallData,
+                    Some(src_addr.expr() + i.expr()),
+                    buffer_get_data.byte(i),
+                )
+            });
+            // Write bytes[i] to memory when selectors[i] != 0
+            cb.condition(buffer_get_data.has_data(i), |cb| {
+                cb.memory_lookup_with_counter(
+                    rw_counter.clone() + rw_counter_offset.clone(),
+                    1.expr(),
+                    dst_addr.expr() + i.expr(),
+                    buffer_get_data.byte(i),
+                )
+            });
+            rw_counter_offset = rw_counter_offset + buffer_get_data.has_data(i);
+        }
+
+        let copied_size = buffer_get_data.num_bytes();
+        let finish_gadget = ComparisonGadget::construct(
+            cb,
+            copied_size.clone(),
+            bytes_left.expr(),
+        );
+        let (lt, finished) = finish_gadget.expr();
+        cb.add_constraint(
+            "Constrain num_bytes <= bytes_left",
+            lt * finished.clone(),
+        );
+
+        // When finished == 0, constraint the CopyToMemory state in next step
+        let next_src_addr = cb.query_cell_next_step();
+        let next_dst_addr = cb.query_cell_next_step();
+        let next_bytes_left = cb.query_cell_next_step();
+        let next_src_addr_end = cb.query_cell_next_step();
+        let next_from_tx = cb.query_cell_next_step();
+        let next_tx_id = cb.query_cell_next_step();
+        cb.condition(1.expr() - finished.clone(), |cb| {
+            cb.require_next_state(ExecutionState::CopyToMemory);
+            cb.add_constraint(
+                "src_addr + copied_size == next_src_addr",
+                src_addr.expr() + copied_size.clone() - next_src_addr.expr(),
+            );
+            cb.add_constraint(
+                "dst_addr + copied_size == next_dst_addr",
+                dst_addr.expr() + copied_size.clone() - next_dst_addr.expr(),
+            );
+            cb.add_constraint(
+                "length == copied_size + next_length",
+                bytes_left.expr()
+                    - copied_size.clone()
+                    - next_bytes_left.expr(),
+            );
+            cb.add_constraint(
+                "src_addr_bound == next_src_addr_bound",
+                src_addr_end.expr() - next_src_addr_end.expr(),
+            );
+            cb.add_constraint(
+                "from_tx == next_from_tx",
+                from_tx.expr() - next_from_tx.expr(),
+            );
+            cb.add_constraint(
+                "tx_id == next_tx_id",
+                tx_id.expr() - next_tx_id.expr(),
+            );
+        });
+
+        // State transition
+        let step_state_transition = StepStateTransition {
+            rw_counter: Delta(rw_counter_offset),
+            program_counter: Delta(finished),
```

Comment:
moving `program_counter` counts into real op code seems make more sense as only real op code will change pc location of evm,  here use default zero without explicitly assignment, what do you think?

Reply:
Changed